processor/spanmetricsprocessor/processor.go

```diff
@@ -384,64 +426,113 @@ func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Sp
 
-	p.lock.Lock()
-	p.cache(serviceName, span, key, resourceAttr)
-	p.updateCallMetrics(key)
-	p.updateLatencyMetrics(key, latencyInMilliseconds, index)
-	p.updateLatencyExemplars(key, latencyInMilliseconds, span.TraceID())
-	p.lock.Unlock()
+	p.cacheMetricKey(span, mKey, resourceAttr)
+	p.cacheResourceAttrKey(serviceName, resourceAttr, resourceAttrKey)
+	p.updateCallMetrics(resourceAttrKey, mKey)
+	p.updateLatencyMetrics(resourceAttrKey, mKey, latencyInMilliseconds, index)
+	p.updateLatencyExemplars(resourceAttrKey, mKey, latencyInMilliseconds, span.TraceID())
 }
 
 // updateCallMetrics increments the call count for the given metric key.
-func (p *processorImp) updateCallMetrics(key metricKey) {
-	p.callSum[key]++
+func (p *processorImp) updateCallMetrics(rKey resourceKey, mKey metricKey) {
+	rcs, ok := p.callSum[rKey]
+
+	if !ok {
+		p.callSum[rKey] = map[metricKey]int64{mKey: 1}
+		return
+	}
+
+	if _, ok := rcs[mKey]; !ok {
+		rcs[mKey] = 1
+		return
+	}
+
+	rcs[mKey]++
+}
+
+func (p *processorImp) reset() {
+	// If delta metrics, reset accumulated data
+	if p.config.GetAggregationTemporality() == pdata.MetricAggregationTemporalityDelta {
+		p.resetAccumulatedMetrics()
+	} else {
+		p.metricKeyToDimensions.RemoveEvictedItems()
+		p.resourceKeyToDimensions.RemoveEvictedItems()
+	}
+
+	p.resetExemplarData()
 }
 
 // resetAccumulatedMetrics resets the internal maps used to store created metric data. Also purge the cache for
 // metricKeyToDimensions.
 func (p *processorImp) resetAccumulatedMetrics() {
-	p.callSum = make(map[metricKey]int64)
-	p.latencyCount = make(map[metricKey]uint64)
-	p.latencySum = make(map[metricKey]float64)
-	p.latencyBucketCounts = make(map[metricKey][]uint64)
+	p.callSum = make(map[resourceKey]map[metricKey]int64)
+	p.latencyCount = make(map[resourceKey]map[metricKey]uint64)
+	p.latencySum = make(map[resourceKey]map[metricKey]float64)
+	p.latencyBucketCounts = make(map[resourceKey]map[metricKey][]uint64)
 	p.metricKeyToDimensions.Purge()
+	p.resourceKeyToDimensions.Purge()
 }
 
-// updateLatencyExemplars sets the histogram exemplars for the given metric key and append the exemplar data.
-func (p *processorImp) updateLatencyExemplars(key metricKey, value float64, traceID pdata.TraceID) {
-	if _, ok := p.latencyExemplarsData[key]; !ok {
-		p.latencyExemplarsData[key] = []exemplarData{}
+// updateLatencyExemplars sets the histogram exemplars for the given resource and metric key and append the exemplar data.
+func (p *processorImp) updateLatencyExemplars(rKey resourceKey, mKey metricKey, value float64, traceID pdata.TraceID) {
+	rled, ok := p.latencyExemplarsData[rKey]
+	if !ok {
+		rled = make(map[metricKey][]exemplarData)
+		p.latencyExemplarsData[rKey] = rled
+	}
+
+	if _, ok = rled[mKey]; !ok {
+		rled[mKey] = []exemplarData{}
 	}
 
 	e := exemplarData{
 		traceID: traceID,
 		value:   value,
 	}
-	p.latencyExemplarsData[key] = append(p.latencyExemplarsData[key], e)
+	rled[mKey] = append(rled[mKey], e)
```

Comment:
Why can't we replace this with the following?:
```
rled[mKey] = append(rled[mKey], exemplarData{traceID, value})
```

I also think `exemplarData{traceID, value}` is more readable, given there's less stutter.

I've written an example to demonstrate: https://go.dev/play/p/fWyh0TOfOZO